myia/opt/lib.py

```diff
@@ -289,13 +289,17 @@ def replace_applicator(optimizer, node, equiv):
 
     For example, `lambda x, y: f(x, y)` is replaced by f.
 
-    The function must be applied on all the parameters in the exact order
-    given.
+    The inner function must be applied on all the outer function's parameters
+    in the exact same order, and it must be either a Primitive or a global
+    function.
     """
     g = equiv[G].value
     out = g.output
     if is_apply(out) and out.inputs[1:] == g.parameters:
-        return out.inputs[0]
+        inner = out.inputs[0]
+        if is_constant(inner, Primitive) \
+                or is_constant_graph(inner) and inner.value.parent is None:
```

Comment:
Leave a comment to note that if parent is not `g` this should be ok, but with `None` we are 100% sure it is ok.

Reply:
I'll do that.